Add check for currently open H5 objects before insert/get

knelli2 opened the pull request:
## Proposed changes

Previously, if you had an object open from an H5 file and you tried to either `insert`/`get` another object, the H5 file would allow you to do that by silently closing the previous object internally. This is correct behavior in the sense that only one object can be open at a time. However, this leads to a very poor error message if you do accidentally try to get a reference to two different objects and then use them both (some low level h5 error about the dataset not existing). This PR adds a check in the `get`, `insert`, and `try_insert` functions of an H5 file to ensure there are no currently open objects. If you forgot to close a previous object before using a new one an ERROR is thrown.

To make the error messages more helpful, the path of the currently open object is given along with the path of the object you are trying to get. This is done by adding a `subfile_path()` function to the `Object` base class.

Also, makes this clearer in the docs of the H5 file by making it a warning. Before, the requirement that only one object could be open at a time was sort of hidden in the middle of the docs.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

When `insert`ing or `get`ing an object in an H5 file, you must `close_current_object()` before you can `get`/`insert` a new object.

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Add check for open H5 objects

knelli2 committed:
Add subfile_path function to H5 Object base class

knelli2 committed:
Store subfile path in OpenGroup

nilsdeppe commented on src/IO/H5/OpenGroup.hpp:
```diff
@@ -49,8 +49,11 @@ class OpenGroup {
 
   const hid_t& id() const { return group_id_; }
 
+  std::string group_path_with_trailing_slash() const { return group_path_str_; }
```
`const std::string& group_...`?

nilsdeppe commented on src/IO/H5/Dat.hpp:
```diff
@@ -118,12 +118,18 @@ class Dat : public h5::Object {
+  /*!
+   * \returns the path to the subfile where this object is stored
+   */
+  std::string subfile_path() const override { return path_; }
```
`const std::string&`?

nilsdeppe commented on src/IO/H5/Header.hpp:
```diff
@@ -52,12 +52,16 @@ class Header : public h5::Object {
   std::string get_build_info() const;
 
+  /// Return the path to the subfile where this object is stored
+  std::string subfile_path() const override { return path_; }
```
`const std::string&`?

nilsdeppe commented on src/IO/H5/Object.hpp:
```diff
@@ -21,5 +21,8 @@ class Object {
   /// \endcond
+
+  /// Return the path to the subfile where this object is stored
+  virtual std::string subfile_path() const = 0;
```
`const std::string&`?

nilsdeppe commented on src/IO/H5/SourceArchive.hpp:
```diff
@@ -34,10 +34,14 @@ class SourceArchive : public h5::Object {
   const std::vector<char>& get_archive() const { return source_archive_; }
 
+  /// Return the path to the subfile where this object is stored
+  std::string subfile_path() const override { return path_; }
```
`const std::string&`?

nilsdeppe commented on src/IO/H5/StellarCollapseEos.hpp:
```diff
@@ -73,8 +73,12 @@ class StellarCollapseEos : public h5::Object {
       const std::string& dataset_name) const;
 
+  /// Return the path to the subfile where this object is stored
+  std::string subfile_path() const override { return path_; }
```
`const std::string&`?

nilsdeppe commented on src/IO/H5/Version.hpp:
```diff
@@ -54,8 +54,12 @@ class Version : public h5::Object {
   uint32_t get_version() const { return version_; }
 
+  /// Return the path to the subfile where this object is stored
+  std::string subfile_path() const override { return path_; }
```
`const std::string&`?

I thought we enabled Doxygen inheritance of parent virtual function documentation. If you want you could check and then remove the dox string from the derived class `subfile_path` function. Up to you!

nilsdeppe commented on src/IO/H5/VolumeData.hpp:
```diff
@@ -152,9 +152,13 @@ class VolumeData : public h5::Object {
       size_t observation_id) const;
 
+  /// Return the path to the subfile where this object is stored
+  std::string subfile_path() const override { return path_; }
```
`const std::string&`?

nilsdeppe commented on src/Domain/FunctionsOfTime/ReadSpecPiecewisePolynomial.cpp:
```diff
@@ -31,6 +31,7 @@ void read_spec_piecewise_polynomial(
     const bool quaternion_rotation) {
   h5::H5File<h5::AccessType::ReadOnly> file{file_name};
   for (const auto& [spec_name, spectre_name] : dataset_name_map) {
+    file.close_current_object();
```
Sorry, shouldn't this be at the end of the for loop?

knelli2 replied:
That isn't the best place. Just below this on line 45-46, there's a `continue`. If the closing was at the end of the loop, it'd also have to be there just before the `continue`, otherwise a file would still be open on the next iteration of the loop (since it didn't make it to the bottom). Having it at the top just before we open the file guarantees that no other objects are open and reduces the amount of places you have to remember to close the file.

nilsdeppe replied:
Okay, fair enough!

nilsdeppe commented on src/IO/H5/File.hpp:
```diff
@@ -204,7 +207,10 @@ template <AccessType Access_t>
 const ObjectType& H5File<Access_t>::get(const std::string& path,
                                         Args&&... args) const {
-  current_object_ = nullptr;
+  if (current_object_ != nullptr) {
```
If you want you might able able to factor this check into a private member function and just call that in the 3 places. That way if the message gets updated it only needs to be done so once

knelli2 replied:
I left them separate because the messages are slightly different depending on the function.
- `get` -> "Object ... already open. Cannot open object ..."
- `insert` -> "Object ... already open. Cannot insert object ..."
- `try_insert` -> "Object ... already open. Cannot try to insert object ..."

I found this helpful when encountering the error to know what action I was trying to do so I could more easily find the offending line.

nilsdeppe replied:
Ah, yea, I agree. Leave as-is :)

nilsvu commented:
@nikwit would you be willing to do the second review on this, or just review the changes to the interpolation script?

knelli2 commented:
Changed all the `const std::string&` and removed the dox from each inherited class function (Doxygen does indeed have inherited classes use the docs of the base class). No other changes.

nilsdeppe commented:
@nikwit do you have time to look at this? If not, that's fine :)

nikwit commented on src/Visualization/Python/InterpolateVolumeData.py:
```diff
@@ -143,6 +147,8 @@ def interpolate_h5_file(source_file_path,
         ]
 
+        source_file.close()
+
```
I found I need to close `target_file` in line 179 before `get_vol` is called for this to work.

nikwit reviewed (commented):
Looks good except one thing. Sorry for the delay.
